pkg/operator/controller/status.go

```diff
@@ -40,71 +46,153 @@ func (r *reconciler) syncOperatorStatus() error {
 	}
 
-	if !operatorStatusesEqual(*oldStatus, co.Status) {
-		err = r.client.Status().Update(context.TODO(), co)
-		if err != nil {
-			return fmt.Errorf("failed to update clusteroperator %s: %v", co.Name, err)
-		}
+	ingressList := &operatorv1.IngressControllerList{}
+	if err := r.client.List(context.TODO(), ingressList, client.InNamespace(r.Namespace)); err != nil {
+		return nil, fmt.Errorf("failed to list IngressControllers: %v", err)
 	}
 
-	return nil
+	return ingressList.Items, nil
 }
 
-// getOperatorState gets and returns the resources necessary to compute the
-// operator's current state.
-func (r *reconciler) getOperatorState() (*corev1.Namespace, []operatorv1.IngressController, error) {
-	ns := manifests.RouterNamespace()
+// computeOperatorStatusVersions computes the operator's current versions and also updates progressing condition if needed.
+func (r *reconciler) computeOperatorStatusVersions(oldVersions []configv1.OperandVersion, conditions []configv1.ClusterOperatorStatusCondition) ([]configv1.OperandVersion, []configv1.ClusterOperatorStatusCondition) {
+	oldReleaseVersion := UnknownVersionValue
+	oldIngressControllerImage := UnknownVersionValue
+	for _, opv := range oldVersions {
+		switch opv.Name {
+		case OperatorVersionName:
+			oldReleaseVersion = opv.Version
+		case IngressControllerVersionName:
+			oldIngressControllerImage = opv.Version
+		}
+	}
 
-	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: ns.Name}, ns); err != nil {
-		if errors.IsNotFound(err) {
-			return nil, nil, nil
+	operatorAvailable := false
+	operatorProgressing := false
+	var progressingCondition *configv1.ClusterOperatorStatusCondition
+	for i, cond := range conditions {
+		switch cond.Type {
+		case configv1.OperatorAvailable:
+			if cond.Status == configv1.ConditionTrue {
+				operatorAvailable = true
+			}
+		case configv1.OperatorProgressing:
+			progressingCondition = &conditions[i]
+			if cond.Status == configv1.ConditionTrue {
+				operatorProgressing = true
+			}
 		}
+	}
 
-		return nil, nil, fmt.Errorf(
-			"error getting Namespace %s: %v", ns.Name, err)
+	releaseVersion := r.OperatorReleaseVersion
+	ingressControllerImage := r.IngressControllerImage
+	// We need to report old version until the operator fully transitions to the new version.
+	// https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/clusteroperator.md#version-reporting-during-an-upgrade
+	if !operatorAvailable || operatorProgressing {
+		releaseVersion = oldReleaseVersion
+		ingressControllerImage = oldIngressControllerImage
 	}
 
-	ingressList := &operatorv1.IngressControllerList{}
-	if err := r.client.List(context.TODO(), ingressList, client.InNamespace(r.Namespace)); err != nil {
-		return nil, nil, fmt.Errorf("failed to list IngressControllers: %v", err)
+	// Update progressing condition if needed
+	if (releaseVersion != r.OperatorReleaseVersion) || (ingressControllerImage != r.IngressControllerImage) {
+		if !operatorProgressing {
```

Comment:
Can you help me understand why progressing is set to true/"Reconciling" when `operatorProgressing` is false? It seems like it would be the opposite.

Reply:
`releaseVersion != r.OperatorReleaseVersion` is true if we did not move to the new version yet. This could happen when the operator condition is not available or is progressing. 
If the operator condition is progressing, progressing condition status and reason is already set and we just need to update the message with version info.
If the operator condition is not progressing (number of ingresses available and desired matched) and not available, then we do need to set progressing condition status to true and the reason as 'reconciling' to reflect the progressing state due to version mismatch.